src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
     fire_and_forget TerminalPage::_PasteFromClipboardHandler(const IInspectable /*sender*/,
                                                              const PasteFromClipboardEventArgs eventArgs)
+    try
     {
-        const auto data = Clipboard::GetContent();
-
-        // This will switch the execution of the function to a background (not
-        // UI) thread. This is IMPORTANT, because the getting the clipboard data
-        // will crash on the UI thread, because the main thread is a STA.
-        co_await winrt::resume_background();
+        // Counterintuitively pasting the clipboard should block the UI thread, because that's
+        // actually the expectation of users: If you type something, paste and then type something again,
+        // text should be written in exactly that order, no matter how loaded the CPU is. See GH#15315.
+        assert(Dispatcher().HasThreadAccess());
 
-        try
+        if (!IsClipboardFormatAvailable(CF_UNICODETEXT))
         {
-            hstring text = L"";
-            if (data.Contains(StandardDataFormats::Text()))
-            {
-                text = co_await data.GetTextAsync();
-            }
-            // Windows Explorer's "Copy address" menu item stores a StorageItem in the clipboard, and no text.
-            else if (data.Contains(StandardDataFormats::StorageItems()))
+            co_return;
+        }
+
+        // The Win32 clipboard API allows us to synchronously retrieve the clipboard on the UI thread
+        // which is exactly what we want (see above). Additionally...
+        //
+        // The old Win32 clipboard API as used below is somewhere in the order of 1000x faster than the WinRT one,
+        // which has about the same latency as a fibre optics connection across the entire WA state.
+        // Don't use the WinRT clipboard API. RPC calls are not funny.
+        winrt::hstring text;
+        {
+            // According to various reports on the internet, OpenClipboard might
+            // fail to acquire the internal lock over RDP, due to rdpclip.exe.
+            // I haven't seen this happen yet myself however.
+            for (int attempts = 0;;)
             {
-                auto items = co_await data.GetStorageItemsAsync();
-                if (items.Size() > 0)
+                if (!OpenClipboard(nullptr))
                 {
-                    auto item = items.GetAt(0);
-                    text = item.Path();
+                    if (++attempts > 5)
+                    {
+                        co_return;
+                    }
+
+                    Sleep(5);
+                    continue;
                 }
+
+                break;
             }
 
-            if (_settings.GlobalSettings().TrimPaste())
+            const auto clipboardCleanup = wil::scope_exit([]() {
+                CloseClipboard();
+            });
+
+            const auto data = GetClipboardData(CF_UNICODETEXT);
```

Comment:
"Copy as path" does immediately set a `CF_UNICODETEXT`, as tested via [InsideClipboard](https://www.nirsoft.net/utils/inside_clipboard.html). But that comment is actually about a way less known feature in Explorer... Namely this context menu on the address bar:
![image](https://github.com/microsoft/terminal/assets/2256941/a0c7a40c-10e0-4e40-bd42-b04ad83d6854)

The first item ("Copy address") sets the clipboard to contain these items:
* `CF_HDROP`
* `FileName`
* `FileNameW`
* `CF_LOCALE`

The clipboard doesn't contain a `CF_UNICODETEXT`, or anything similar at all. I do not understand how the translation to a `CF_UNICODETEXT` works in this case, but in my testing, I found that it works incredibly consistent. I suspect this might be because this Explorer functionality is incredibly old at this point (= was written around the time the win32 clipboard API was still maintained).

Reply:
My guess is the system (user32.dll or win32kfull.sys) can take CF_HDROP and synthesize CF_WhateverTEXT, at least when just one file is referenced in the CF_HDROP.

It's OK to close this thread out; it's at your discretion whether to add the comment I suggested in point 1.